File: ruby/rna-transcription/rna_transcription.rb
module BookKeeping
  VERSION = 4
end

class Complement
  DNAtoRNA = {
    "G" => "C",
    "C" => "G",
    "T" => "A",
    "A" => "U",
  }

  class InvalidBase < RuntimeError
  end

  # DNAtoRNA.default_proc = -> (hash, key) { raise InvalidBase, "#{key} is not a DNA base!" }

  VERSION = 1
  def self.of_dna(strand)
    strand.split("")
      .map { |dna| DNAtoRNA[dna] || (break []) }
      .join

    # rescue InvalidBase
    # ""
  end
end

# Given a DNA strand, its transcribed RNA strand is formed by replacing
# each nucleotide with its complement:

# * `G` -> `C`
# * `C` -> `G`
# * `T` -> `A`
# * `A` -> `U`

File: python/rna-transcription/rna_transcription.py
def to_rna(dna_strand):
    DNAtoRNA = {'C': 'G', 'A': 'U', 'T': 'A', 'G': 'C'}
    rna = list(map(lambda x: DNAtoRNA[x], dna_strand))
    return "".join(rna)
